internal/controller/csiaddons/persistentvolumeclaim_controller.go

```diff
@@ -146,59 +153,59 @@ func (r *PersistentVolumeClaimReconciler) Reconcile(ctx context.Context, req ctr
-// returns false and if the driver supports space reclamation, it returns true.
-// If the driver name is not registered in the connection pool, it returns
-// false and requeues the request.
-func (r *PersistentVolumeClaimReconciler) checkDriverSupportReclaimsSpace(logger *logr.Logger, annotations map[string]string, driver string) (bool, bool) {
-	reclaimSpaceSupportedByDriver := false
-
-	if drivers, ok := annotations[rsCSIAddonsDriverAnnotation]; ok && slices.Contains(strings.Split(drivers, ","), driver) {
-		reclaimSpaceSupportedByDriver = true
-	}
-
-	ok := r.supportsReclaimSpace(driver)
-	if reclaimSpaceSupportedByDriver && !ok {
-		logger.Info("Driver supports spacereclamation but driver is not registered in the connection pool, Reqeueing request", "DriverName", driver)
-		return true, false
-	}
-
-	if !ok {
-		logger.Info("Driver does not support spacereclamation, skip Requeue", "DriverName", driver)
+// checkDriverSupportCapability checks if the given driver supports the specified capability.
+// It is also used to determine the requeue in case the driver supports the capability
+// but the capability is not found in connection pool.
+// The first return value is true when a requeue is needed.
+// The second return value is true when the capability is registered.
+func (r *PersistentVolumeClaimReconciler) checkDriverSupportCapability(
+	logger *logr.Logger,
+	annotations map[string]string,
+	driverName string,
+	cap Operation) (bool, bool) {
+	driverSupportsCap := false
+	capFound := false
+
+	var driverAnnotation string
+	switch cap {
+	case relciamSpaceOp:
+		driverAnnotation = rsCSIAddonsDriverAnnotation
+	case keyRotationOp:
+		driverAnnotation = krCSIAddonsDriverAnnotation
+	default:
+		logger.Info("Unknown capability", "Capability", cap)
 		return false, false
 	}
 
-	return false, true
-}
-
-// checkDriverSupportsEncryptionKeyRotate checks if the driver supports key
-// rotation or not. If the driver does not support key rotation, it
-// returns false and if the driver supports key rotation, it returns true.
-// If the driver name is not registered in the connection pool, it returns
-// false and requeues the request.
-func (r *PersistentVolumeClaimReconciler) checkDriverSupportsEncryptionKeyRotate(
-	logger *logr.Logger,
-	annotations map[string]string,
-	driver string) (bool, bool) {
-	keyRotationSupportedByDriver := false
+	if drivers, ok := annotations[driverAnnotation]; ok && slices.Contains(strings.Split(drivers, ","), driverAnnotation) {
+		driverSupportsCap = true
+	}
 
-	if drivers, ok := annotations[krCSIAddonsDriverAnnotation]; ok && slices.Contains(strings.Split(drivers, ","), driver) {
-		keyRotationSupportedByDriver = true
+	conns := r.ConnPool.GetByNodeID(driverName, "")
+	for _, conn := range conns {
+		for _, c := range conn.Capabilities {
+			switch cap {
+			case relciamSpaceOp:
+				capFound = c.GetReclaimSpace() != nil
+			case keyRotationOp:
+				capFound = c.GetEncryptionKeyRotation() != nil
+			}
+		}
```

Comment:
```suggestion
			default:
				continue
			}
			break
		}
		if capFound{
		         break
		}
```

We will have to exist after finding the cap otherwise it will get overwritten ?

Reply:
updated the comment, refer to  it again.

Reply:
Right, overlooked the fact that a connection can have multiple capabilities.